src/Workspaces/Core/Portable/Storage/SQLite/v2/Interop/SqlStatement.cs

```diff
@@ -67,7 +69,34 @@ public Result Step(bool throwOnError = true)
         }
 
         internal void BindStringParameter(int parameterIndex, string value)
-            => _connection.ThrowIfNotOk(NativeMethods.sqlite3_bind_text(_rawStatement, parameterIndex, value));
+        {
+            // Attempt to stackalloc utf8 converted small strings to avoid lots of allocs.
+            var utf8ByteCount = Encoding.UTF8.GetByteCount(value);
+            var utf8WithTrailingZeroByteCount = utf8ByteCount + 1;
+
+            if (utf8WithTrailingZeroByteCount <= 512)
+            {
+                Span<byte> bytes = stackalloc byte[utf8WithTrailingZeroByteCount];
+#if NETCOREAPP
+                Contract.ThrowIfFalse(Encoding.UTF8.GetBytes(value.AsSpan(), bytes) == utf8ByteCount);
+#else
+                unsafe
+                {
+                    fixed (char* charsPtr = value)
+                    fixed (byte* bytesPtr = bytes)
+                    {
+                        Contract.ThrowIfFalse(Encoding.UTF8.GetBytes(charsPtr, value.Length, bytesPtr, utf8WithTrailingZeroByteCount) == utf8ByteCount);
+                    }
+                }
+#endif
+                bytes[utf8WithTrailingZeroByteCount - 1] = 0;
+                _connection.ThrowIfNotOk(NativeMethods.sqlite3_bind_text(_rawStatement, parameterIndex, bytes));
```

Comment:
Oh, hang on. You're giving native code a pointer to utf8 bytes that are on the stack. When your bind method returns, guess what bytes now get stomped on, leading to corrupted data being used by your query!

Reply:
is 'bind' not copying the data into itself?  @ericsink ?

If it's not doing this then i have *no* idea how this is possibly working and not dying a horrible death in CI.

Reply:
The actual version of bind has a param to specify whether to copy or not. I don't recall what my code is doing with that. I need to check. It would be nice to have sqlitepclraw expose the full flexibility of the underlying C API.

Reply:
Would def appreciate any information you have here @ericsink. If you are copying, then the approach I'm taking off appropriate and will help avoid an unnecessary copy.

Reply:
The native code offers `sqlite3_bind_text16` which would make the whole thing moot as encoding wouldn't be necessary.

Reply:
The PCL raw does _not_ pass in `SQLITE_TRANSIENT` which is [what causes sqlite to copy the value](https://sqlite.org/c3ref/bind_blob.html). It passes in -1, which means the length is determined by the first null character:

https://github.com/ericsink/SQLitePCL.raw/blob/ced6c13984f604f6d91d27f1c0f46a06ada46e37/src/SQLitePCLRaw.provider.dynamic_cdecl/Generated/provider_dynamic_cdecl.cs#L1181

Reply:
Oh, I got that wrong. It does pass in the -1 as the last parameter, but the length parameter is set to the encoded length. I don't know which special constant -1 maps to.

Reply:
The special constant -1 is `SQLITE_TRANSIENT`.

Reply:
So in other words, I'm telling SQLite to make its own copy of the string.  The other options are to have it retain the pointer and treat it as static, or to provide a callback to destroy the string.  Neither of those seem at all suitable for use with stack allocated memory.

Reply:
Apparently [-1 means SQLITE_TRANSIENT](https://stackoverflow.com/a/26884081/46926), so indeed a copy is immediate and your code should be fine.
Blob binding similarly copies the value.

Reply:
I believe that using `sqlite3_bind_text16` would merely change where the the encoding conversion happens.  SQLite will convert it to utf8.

Reply:
Well, having sqlite do the encoding sounds easier than what Cyrus is doing here. IMO it makes sense to expose it, or better yet just use it underneath the existing `string` overload to avoid the GH heap allocations.

Reply:
> better yet just use it underneath the existing string overload to avoid the GH heap allocations.

This sounds like the real fix.

Reply:
> This sounds like the real fix, and the API could support both string and ReadOnlySpan<char> efficiently.

Sure.  I'll see about making a PR for sqlite on this.